### DISCUSSION TITLE
Feature #1.1 - Registration and Log In/Out

### DISCUSSION THREAD

**Discussion by Nyero2023** [ASCENDynamics-NFP/AscendCoopPlatform#140] [Epic 1: User Registration and Authentication Flow]
### Description
This feature manages the entry point for users, allowing them to register and access the platform. It involves creating intuitive and secure registration and login pages for our users. These pages should provide a straightforward way for users to create an account or access their existing one.
#9 

### Story
Create an Account or Registration

**User Story:** **As a** User, **I want to** be able to create a new account **so that** I can access the platform's features and connect with other users.

**Acceptance Criteria**
- Users should be able to see a clear and inviting "Sign Up" or "Create Account" button.
- The user clicks on the "Sign Up" button, and then sees a screen that prompts the identification of the user as either "Group" or "User." 
- The user selects one of the options either "Group" or "User" and is directed to a registration page where they enter their basic information, as follows:
  - **"Group"**
    - **Section 1:**
      - Email Address
      - Password
        - Password requirements are shown.
          1. At least 8 character limit.
          2. The character should have the following: uppercase, lowercase, numeric, or special characters. The allowed special characters are ~! @ # $ % ^ * - _ = + [ { ] } / ; : , . ?
      - Confirm Password
    - **Section 2:**
      - **Basic Information**
        - Group Name _(Required)_
        - Group Type _(e.g., Nonprofit, Community Group, Interest-Based Group)_
        - Group Tagline _(Required)_
        - Group Description _(Brief overview of the group’s purpose, activities, and goals)_
        - Group Logo/Image _(Optional)_ (Images will only be able to get added once the group is created)
        - Group Header/Image _(Optional)_ (Images will only be able to get added once the group is created)

      - **Contact & Location**
        - Contact Email _(Required)_
        - Contact Number _(Optional)_
        - Website URL _(Optional)_
        - Physical Address _(Optional)_
        - City/Region _(Optional)_
        - Country _(Optional)_

  - **"User"**
    - **Section 1**
      - Email Address
      - Password
        - Password requirements are shown.
          1. At least 8 character limit.
          2. The character should have the following: uppercase, lowercase, numeric, or special characters. The allowed special characters are ~! @ # $ % ^ * - _ = + [ { ] } / ; : , . ?
      - Confirm Password
    - **Section 2**
      - **Contact Information**
        - Phone Number
        - Alternate Email _(Optional)_
        - Mailing Address _(Optional)_
        - Emergency Contact Number _(Optional)_
        - Preferred Method of Contact _(e.g., Email, Phone, Text)_

      - **Professional Information**
        - Occupation
        - Employer Name _(Optional)_
        - Work Experience _(Optional)_
        - Skills and Expertise _(Optional)_
        - Current Job Title _(Optional)_
        - LinkedIn Profile _(Optional)_
        - Educational Background _(Optional)_

- The user should also have the option to sign up using Single Sign-On (SSO).

### Story
Develop a Log In/Out functionality
**User Story:** As a user, I want a seamless and secure experience while logging into and out of my account so that I can control access to my account.

**Acceptance Criteria**
Log In
- Users should be able to access the login page from the platform's home screen.
- The login page should prominently display input fields for entering users' email/username and password.
- The system should validate users' credentials, ensuring that the entered email/username is registered and the password is correct.
- If the credentials are valid the system should navigate me to my personalized dashboard.
- In case of invalid credentials, the system should display a clear error message, guiding me to correct and re-enter the information.

Log Out
- While logged into my account, there should be a visible and easily accessible "Log Out" option.
- Upon confirmation, the system should securely log me out, terminating my active session and redirecting me to the login page.
- After logging out, attempting to access any authenticated page should redirect me to the login page.

**Edge Cases**
Edge Cases for Registration
- Duplicate Email/Username:
  - User attempts to register with an email/username that already exists in the system.
  - The system should provide a clear error message prompting the user to choose a different email/username.
- Invalid Email Format:
  - User enters an email address with an incorrect format during registration.
  - The system should validate the email format and provide a helpful error message guiding the user to enter a valid email address.
- Weak Password:
  - User tries to register with a password that does not meet the platform's security requirements.
  - The system should enforce password strength rules and communicate them clearly to the user.
- Incomplete Registration Form:
  - User submits the registration form with missing or incomplete mandatory fields.
  - The system should highlight the missing fields and guide the user to complete the form before proceeding.

Edge Cases for Log In/Out
- Forgotten Password:
  - User forgets their password and initiates a password reset.
  - The system should guide the user through a secure password reset process, verifying their identity through email or other means.
- Concurrent Sessions:
  - User attempts to log in from multiple devices simultaneously.
  - The system should handle concurrent sessions according to security policies, either allowing or restricting simultaneous logins.

**Comment by Nyero2023** [comment]
### Tasks
Here is a link to the list of tasks defined, #144

**Comment by mufniarz** [comment]
For users we would like to collect this information: https://docs.google.com/document/d/1FeEtxbx20-ITCe_7-Y5vuspiQ1vQnZZA5PWYajthWVY/edit?usp=sharing
For groups we would like to collect this information: https://docs.google.com/document/d/17jjfrS_wFh8mUuJpAJx8Hhi8ol6ja5IpjpziZ8Lk-CE/edit?usp=sharing
